File: api/services/interfaces.py
"""Segregated service interfaces following ISP - Dependency Inversion Principle."""
from abc import ABC, abstractmethod
from typing import TypeVar, Generic, List, Optional

T = abstractmethod  # Type for read operations
U = abstractmethod  # Type for write operations (DTOs)


class IReadService(ABC, Generic[T]):
    """Interface for read-only service operations."""

    @abstractmethod
    def get_by_id(self, id: int) -> Optional[T]:
        """Retrieve a single item by ID."""
        pass

    @abstractmethod
    def get_all(self) -> List[T]:
        """Retrieve all items."""
        pass


class ISearchService(ABC, Generic[T]):
    """Interface for search/filter operations."""

    @abstractmethod
    def search(self, **filters) -> List[T]:
        """Search items by filters."""
        pass

    @abstractmethod
    def count(self, **filters) -> int:
        """Count items matching filters."""
        pass


class ICreateService(ABC, Generic[U, T]):
    """Interface for create operations - clients depend only on what they create."""

    @abstractmethod
    def create(self, data: U) -> T:
        """Create a new item."""
        pass


class IUpdateService(ABC, Generic[U, T]):
    """Interface for update operations - clients depend only on what they update."""

    @abstractmethod
    def update(self, id: int, data: U) -> Optional[T]:
        """Update an existing item."""
        pass


class IDeleteService(ABC):
    """Interface for delete operations."""

    @abstractmethod
    def delete(self, id: int) -> bool:
        """Delete an item by ID."""
        pass


class ICRUDService(
    IReadService[T],
    ISearchService[T],
    ICreateService[U, T],
    IUpdateService[U, T],
    IDeleteService
):
    """Complete service interface for full CRUD operations."""
    pass

File: client/src/app/services/core/interfaces.ts
/**
 * Segregated Service Interfaces following Interface Segregation Principle (ISP)
 * Each interface represents a single responsibility
 */

import { Observable } from 'rxjs';
import { ApiResponse } from '@app/models/core/APIResponse';
import { WithAuditFields } from '@app/models/core/Auditable';

/**
 * Interface for read operations only
 * Used by components that only need to retrieve data
 */
export interface IReadService<T> {
  getAll(): Observable<ApiResponse<WithAuditFields<T>[]>>;
  getById(id: number): Observable<ApiResponse<WithAuditFields<T>>>;
}

/**
 * Interface for create operations only
 * Used by components that only need to create data
 */
export interface ICreateService<T> {
  create(item: T): Observable<ApiResponse<WithAuditFields<T>>>;
}

/**
 * Interface for update operations only
 * Used by components that only need to update data
 */
export interface IUpdateService<T> {
  update(id: number, item: T): Observable<ApiResponse<WithAuditFields<T>>>;
}

/**
 * Interface for delete operations only
 * Used by components that only need to delete data
 */
export interface IDeleteService {
  delete(id: number): Observable<ApiResponse<any>>;
}

/**
 * Combined CRUD interface for services that need all operations
 */
export interface ICrudService<T>
  extends IReadService<T>,
    ICreateService<T>,
    IUpdateService<T>,
    IDeleteService {}
